substrate/client/network/src/litep2p/ipfs_dht.rs

```diff
@@ -0,0 +1,218 @@
+/// records.
+///
+/// As of November 2025 there are 6.5k reachable IPFS peers with maximum 250M CIDs. This means on
+/// average 250M / (6.5k / 20) ~= 800k CIDs per peer.
+const MAX_PROVIDER_KEYS: usize = 2_000_000;
+
+/// Raw codec type.
+// TODO: index codec along with transaction data and use it instead of the hardcoded one.
+const RAW_CODEC: u64 = 0x55;
+
+/// Interval of Kademlia random walks. Needed to keep the routing table "warm".
+const RANDOM_WALK_INTERVAL: Duration = Duration::from_secs(10 * 60);
+
+/// Maximum allowed number of in-flight Kademlia queries. We need this limit for the networking
+/// stack to not get overwhelmed by unlimited amount of simultaneous `ADD_PROVIDER` queries.
+///
+/// For a maximum 200-second `ADD_PROVIDER` query duration and 100 in-flight queries we have on
+/// average 2 seconds spent on publishing a single provider or 1/3 of the block time. This means
+/// in practice we will spend about one week catching up on transactions in the past two weeks,
+/// before starting publishing new transactions, if we have on average one indexed transaction
+/// per block. So, one transaction per block is a practical limit on the number of indexed
+/// transactions published to IPFS.
+const MAX_INFLIGHT_QUERIES: usize = 100;
+
+/// CID with allocated size of 32 bytes for hash value.
+type Cid = cid::CidGeneric<32>;
+
+pub(crate) struct IpfsDht {
+	kademlia_handle: KademliaHandle,
+	block_provider: Box<dyn BlockProvider>,
+}
+
+impl IpfsDht {
+	pub fn new(
+		bootnodes: Vec<MultiaddrWithPeerId>,
+		block_provider: Box<dyn BlockProvider>,
+	) -> (Self, Config) {
+		let known_peers = {
+			// Convert bootnodes into expected format deduplicating the addresses.
+			let mut known_peers: HashMap<PeerId, HashSet<Multiaddr>> = HashMap::new();
+
+			for address in bootnodes {
+				let peer_id = address.peer_id.into();
+				let multiaddr = address.concat().into();
+
+				known_peers.entry(peer_id).or_default().insert(multiaddr);
+			}
+
+			known_peers.into_iter().map(|(k, v)| (k, v.into_iter().collect())).collect()
+		};
+
+		let (config, kademlia_handle) = ConfigBuilder::new()
+			.with_protocol_names(vec![KAD_PROTOCOL.into()])
+			.with_known_peers(known_peers)
+			.with_max_provider_keys(MAX_PROVIDER_KEYS)
+			.build();
+
+		(Self { kademlia_handle, block_provider }, config)
+	}
+
+	pub async fn run(mut self) {
+		let mut changes = self.block_provider.changes();
+		let mut inflight_queries = HashMap::new();
+		let mut random_walk_query = None;
+		let mut random_walk_interval = tokio::time::interval(RANDOM_WALK_INTERVAL);
+		random_walk_interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+		loop {
+			tokio::select! {
+				change = changes.next(), if inflight_queries.len() < MAX_INFLIGHT_QUERIES => {
+					match change {
+						None => {
+							debug!(target: LOG_TARGET, "BlockProvider terminated, terminating IpfsDht");
+							return
```

Comment:
nit: Should we exit sooner here regardless of `if inflight_queries.len() < MAX_INFLIGHT_QUERIES`?